music_assistant/providers/listenbrainz_scrobble/__init__.py

```diff
@@ -0,0 +1,173 @@
+            self.logger.info("No auth token available, don't forget to configure!")
+            return
+
+        self._client = _get_client(self._get_config())
+
+        # subscribe to internal event
+        self._on_unload.append(
+            self.mass.subscribe(self._on_mass_media_item_played, EventType.MEDIA_ITEM_PLAYED)
+        )
+
+    async def unload(self, is_removed: bool = False) -> None:
+        """
+        Handle unload/close of the provider.
+
+        Called when provider is deregistered (e.g. MA exiting or config reloading).
+        """
+        for unload_cb in self._on_unload:
+            unload_cb()
+
+    async def _on_mass_media_item_played(self, event: MassEvent) -> None:
+        """Media item has finished playing, we'll scrobble the track."""
+        if self._client is None:
+            self.logger.error("no client available during _on_mass_media_item_played")
+            return
+
+        report = event.data
+
+        def make_listen(report: Any) -> Listen:
+            # album artist and track number are not available without an extra API call
+            # so they won't be scrobbled
+
+            # https://pylistenbrainz.readthedocs.io/en/latest/api_ref.html#class-listen
+            return Listen(
+                track_name=report.name,
+                artist_name=report.artist,
+                release_name=report.album,
+                recording_mbid=report.mbid,
+            )
+
+        def update_now_playing() -> None:
+            try:
+                listen = make_listen(report)
+                self._client.submit_playing_now(listen)
+                self.logger.debug(f"track {report.uri} marked as 'now playing'")
+                self._currently_playing = report.uri
+            except Exception as err:
+                self.logger.exception(err)
+
+        def scrobble() -> None:
+            try:
+                listen = make_listen(report)
+                listen.listened_at = int(time.time())
+                self._client.submit_single_listen(listen)
+            except Exception as err:
+                self.logger.exception(err)
+
+        # update now playing if needed
+        if self._currently_playing is None or self._currently_playing != report.uri:
+            await asyncio.to_thread(update_now_playing)
+
+        if self.should_scrobble(report):
+            await asyncio.to_thread(scrobble)
+
+        if report.fully_played:
+            # reset currently playing to avoid it expiring when looping songs
+            self._currently_playing = None
+
+    def should_scrobble(self, report: MediaItemPlaybackProgressReport) -> bool:
+        """Determine if a track should be scrobbled, to be extended later."""
+        # ideally we want more precise control
+        # but because the event is triggered every 30s
+        # and we don't have full queue details to determine
+        # the exact context in which the event was fired
+        # we can only rely on fully_played for now
```

Comment:
you can also rely on the "is_playing" attribute to determine if the track stopped playing or is still ongoing.

Reply:
This came from the last.fm scrobbler. I think they probably should both behave the same?

Maybe at some point it would be interesting to factor out the common bits of scrobbling into a base class or mixin or something.